In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('OnlineRetail.csv', encoding='ISO-8859-1')

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


Nhận xét về dữ liệu:
- Cấu trúc:
  + Bộ dữ liệu có 8 cột và 541909 dòng
  + Các cột InvoiceNo, StockCode, Description, Country thuộc kiểu dữ liệu object
  + Các cột UnitPrice, CustomerID thuộc kiểu dữ liệu float64
  + Cột Quantity thuộc kiễu dữ liệu int64
  * Thuộc tính, thước đo: 
    + InvoiceNo, StockCode, Description, Country là thuốc tính định tính, có thước đo định danh
    + CustomerID tuy mang kiểu dữ liệu là float64 nhưng nó dùng để phân biệt giữ các thực thể khác nhau nên mang thuộc tính định danh và thước đo định danh
    + Quantity, UnitPrice mang thuộc tính định lượng và thước đo ratio

In [5]:
data1 = data.pivot_table(index=['InvoiceNo','Description','Country'], values='Quantity',aggfunc='mean')
data1

Quantity
InvoiceNo Description                         Country                 
536365    CREAM CUPID HEARTS COAT HANGER      United Kingdom       8.0
          GLASS STAR FROSTED T-LIGHT HOLDER   United Kingdom       6.0
          KNITTED UNION FLAG HOT WATER BOTTLE United Kingdom       6.0
          RED WOOLLY HOTTIE WHITE HEART.      United Kingdom       6.0
          SET 7 BABUSHKA NESTING BOXES        United Kingdom       2.0
...                                                                ...
C581490   ZINC T-LIGHT HOLDER STARS SMALL     United Kingdom     -11.0
C581499   Manual                              United Kingdom      -1.0
C581568   VICTORIAN SEWING BOX LARGE          United Kingdom      -5.0
C581569   36 PENCILS TUBE RED RETROSPOT       United Kingdom      -5.0
          HANGING HEART JAR T-LIGHT HOLDER    United Kingdom      -1.0

[529586 rows x 1 columns]

In [8]:
data2 =data.pivot_table(index=['CustomerID'],columns=['StockCode'], values='Quantity', aggfunc={'Quantity':[max,min]})
data2

max                                                           \
StockCode  10002 10080 10120 10123C 10124A 10124G 10125 10133 10135 11001   
CustomerID                                                                  
12346.0      NaN   NaN   NaN    NaN    NaN    NaN   NaN   NaN   NaN   NaN   
12347.0      NaN   NaN   NaN    NaN    NaN    NaN   NaN   NaN   NaN   NaN   
12348.0      NaN   NaN   NaN    NaN    NaN    NaN   NaN   NaN   NaN   NaN   
12349.0      NaN   NaN   NaN    NaN    NaN    NaN   NaN   NaN   NaN   NaN   
12350.0      NaN   NaN   NaN    NaN    NaN    NaN   NaN   NaN   NaN   NaN   
...          ...   ...   ...    ...    ...    ...   ...   ...   ...   ...   
18280.0      NaN   NaN   NaN    NaN    NaN    NaN   NaN   NaN   NaN   NaN   
18281.0      NaN   NaN   NaN    NaN    NaN    NaN   NaN   NaN   NaN   NaN   
18282.0      NaN   NaN   NaN    NaN    NaN    NaN   NaN   NaN   NaN   NaN   
18283.0      NaN   NaN   NaN    NaN    NaN    NaN   NaN   NaN   NaN   NaN   
18287.0      NaN   NaN   NaN    NaN    NaN    NaN   NaN   NaN   NaN   NaN   

            ...    min                                                      
StockCode   ... 90214Y 90214Z BANK CHARGES  C2 CRUK   D DOT    M PADS POST  
CustomerID  ...                                                             
12346.0     ...    NaN    NaN          NaN NaN  NaN NaN NaN  NaN  NaN  NaN  
12347.0     ...    NaN    NaN          NaN NaN  NaN NaN NaN  NaN  NaN  NaN  
12348.0     ...    NaN    NaN          NaN NaN  NaN NaN NaN  NaN  NaN  1.0  
12349.0     ...    NaN    NaN          NaN NaN  NaN NaN NaN  NaN  NaN  1.0  
12350.0     ...    NaN    NaN          NaN NaN  NaN NaN NaN  NaN  NaN  1.0  
...         ...    ...    ...          ...  ..  ...  ..  ..  ...  ...  ...  
18280.0     ...    NaN    NaN          NaN NaN  NaN NaN NaN  NaN  NaN  NaN  
18281.0     ...    NaN    NaN          NaN NaN  NaN NaN NaN  NaN  NaN  NaN  
18282.0     ...    NaN    NaN          NaN NaN  NaN NaN NaN  NaN  NaN  NaN  
18283.0     ...    NaN    NaN          NaN NaN  NaN NaN NaN  1.0  NaN  NaN  
18287.0     ...    NaN    NaN          NaN NaN  NaN NaN NaN  NaN  NaN  NaN  

[4372 rows x 7368 columns]

In [14]:
data3 = data.pivot_table(values=['Quantity','UnitPrice'],index=['StockCode'],columns=['Description'], aggfunc={'Quantity':'sum','UnitPrice':'mean'})

In [15]:
data3

Quantity                                 \
Description   4 PURPLE FLOCK DINNER CANDLES  50'S CHRISTMAS GIFT BAG LARGE   
StockCode                                                                    
10002                                   NaN                            NaN   
10080                                   NaN                            NaN   
10120                                   NaN                            NaN   
10123C                                  NaN                            NaN   
10124A                                  NaN                            NaN   
...                                     ...                            ...   
gift_0001_20                            NaN                            NaN   
gift_0001_30                            NaN                            NaN   
gift_0001_40                            NaN                            NaN   
gift_0001_50                            NaN                            NaN   
m                                       NaN                            NaN   

                                                              \
Description   DOLLY GIRL BEAKER  I LOVE LONDON MINI BACKPACK   
StockCode                                                      
10002                       NaN                          NaN   
10080                       NaN                          NaN   
10120                       NaN                          NaN   
10123C                      NaN                          NaN   
10124A                      NaN                          NaN   
...                         ...                          ...   
gift_0001_20                NaN                          NaN   
gift_0001_30                NaN                          NaN   
gift_0001_40                NaN                          NaN   
gift_0001_50                NaN                          NaN   
m                           NaN                          NaN   

                                                                    \
Description   I LOVE LONDON MINI RUCKSACK  NINE DRAWER OFFICE TIDY   
StockCode                                                            
10002                                 NaN                      NaN   
10080                                 NaN                      NaN   
10120                                 NaN                      NaN   
10123C                                NaN                      NaN   
10124A                                NaN                      NaN   
...                                   ...                      ...   
gift_0001_20                          NaN                      NaN   
gift_0001_30                          NaN                      NaN   
gift_0001_40                          NaN                      NaN   
gift_0001_50                          NaN                      NaN   
m                                     NaN                      NaN   

                                                                   \
Description   OVAL WALL MIRROR DIAMANTE   RED SPOT GIFT BAG LARGE   
StockCode                                                           
10002                                NaN                      NaN   
10080                                NaN                      NaN   
10120                                NaN                      NaN   
10123C                               NaN                      NaN   
10124A                               NaN                      NaN   
...                                  ...                      ...   
gift_0001_20                         NaN                      NaN   
gift_0001_30                         NaN                      NaN   
gift_0001_40                         NaN                      NaN   
gift_0001_50                         NaN                      NaN   
m                                    NaN                      NaN   

                                                                       ...  \
Description   SET 2

In [16]:
data4 = data.pivot_table(values='Quantity',index='InvoiceDate',columns='Description',aggfunc={'Quantity':'sum'})

In [17]:
data4

Description       4 PURPLE FLOCK DINNER CANDLES  \
InvoiceDate                                       
1/10/2011 10:04                             NaN   
1/10/2011 10:07                             NaN   
1/10/2011 10:32                             NaN   
1/10/2011 10:35                             NaN   
1/10/2011 10:36                             NaN   
...                                         ...   
9/9/2011 8:48                               NaN   
9/9/2011 9:03                               NaN   
9/9/2011 9:13                               NaN   
9/9/2011 9:38                               NaN   
9/9/2011 9:52                               NaN   

Description       50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
InvoiceDate                                                           
1/10/2011 10:04                             NaN                 NaN   
1/10/2011 10:07                             NaN                 NaN   
1/10/2011 10:32                             NaN                 NaN   
1/10/2011 10:35                             NaN                 NaN   
1/10/2011 10:36                             NaN                 NaN   
...                                         ...                 ...   
9/9/2011 8:48                               NaN                 NaN   
9/9/2011 9:03                               NaN                 NaN   
9/9/2011 9:13                               NaN                 NaN   
9/9/2011 9:38                               NaN                 NaN   
9/9/2011 9:52                               NaN                 NaN   

Description       I LOVE LONDON MINI BACKPACK   I LOVE LONDON MINI RUCKSACK  \
InvoiceDate                                                                   
1/10/2011 10:04                           NaN                           NaN   
1/10/2011 10:07                           NaN                           NaN   
1/10/2011 10:32                           NaN                           NaN   
1/10/2011 10:35                           NaN                           NaN   
1/10/2011 10:36                           NaN                           NaN   
...                                       ...                           ...   
9/9/2011 8:48                             NaN                           NaN   
9/9/2011 9:03                             NaN                           NaN   
9/9/2011 9:13                             NaN                           NaN   
9/9/2011 9:38                             NaN                           NaN   
9/9/2011 9:52                             NaN                           NaN   

Description       NINE DRAWER OFFICE TIDY   OVAL WALL MIRROR DIAMANTE   \
InvoiceDate                                                              
1/10/2011 10:04                       NaN                          NaN   
1/10/2011 10:07                       NaN                          NaN   
1/10/2011 10:32                       NaN                          NaN   
1/10/2011 10:35                       NaN                          NaN   
1/10/2011 10:36                       NaN                          NaN   
...                                   ...                          ...   
9/9/2011 8:48                         NaN                          NaN   
9/9/2011 9:03                         NaN                          NaN   
9/9/2011 9:13                         NaN                          NaN   
9/9/2011 9:38                         NaN                          NaN   
9/9/2011 9:52                         NaN                          NaN   

Description       RED SPOT GIFT BAG LARGE   SET 2 TEA TOWELS I LOVE LONDON   \
InvoiceDate                                                                   
1/10/2011 10:04                       NaN                               NaN   
1/10/2011 10:07                       NaN                               NaN   
1/10/2011 10:32                       NaN                               NaN   
1/10/2011 10:35           